### PR TITLE
Release/1.1.0

### PR DESCRIPTION
**Pull request by turegjorup**
Map teams for team leads

### CODE REVIEW

File: Migrations/Version20240701112200.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        ADD
+          dept_ids JSON NOT NULL COMMENT \'(DC2Type:json)\',
+        CHANGE
+          company_id company_id INT DEFAULT NULL,
+        CHANGE
+          company_name company_name VARCHAR(255) DEFAULT NULL,
+        CHANGE
+          division_id division_id INT DEFAULT NULL,
+        CHANGE
+          division_name division_name VARCHAR(255) DEFAULT NULL,
```
Comment (rimi-itk) [inline Migrations/Version20240701112200.php:37]: Just out of curiosity: Did you format this yourself or can the `make:migration` command do this? It greatly improves the readability.

File: Service/SamlDTO.php [hunk 2]
```diff
@@ -118,14 +120,161 @@ public function __construct(array $samlAttributes)
...
+            throw new AakSamlException(sprintf('No organization ids given in claims: "%s"', $this->getAttributeValue(self::ID_ARRAY_ATTRIBUTE, $samlAttributes)));
+        }
+
+        // We can map claims from level 1-5. If we see id's for deeper levels we don't know the corresponding claim for
+        // department name.
+        $this->companyId = array_key_exists(0, $this->departmentIds) ? $this->departmentIds[0] : null;
+        $this->divisionId = array_key_exists(1, $this->departmentIds) ? $this->departmentIds[1] : null;
+        $this->departmentId = array_key_exists(2, $this->departmentIds) ? $this->departmentIds[2] : null;
+        $this->subDepartmentId = array_key_exists(3, $this->departmentIds) ? $this->departmentIds[3] : null;
+        $this->officeId = array_key_exists(4, $this->departmentIds) ? $this->departmentIds[4] : null;
```
Comment (rimi-itk) [inline Service/SamlDTO.php]: Any specific reason for not writing this as

```suggestion
        $this->companyId = $this->departmentIds[0] ?? null;
        $this->divisionId = $this->departmentIds[1] ?? null;
        $this->departmentId = $this->departmentIds[2] ?? null;
        $this->subDepartmentId = $this->departmentIds[3] ?? null;
        $this->officeId = $this->departmentIds[4] ?? null;
```

?

File: Service/SamlDTO.php [hunk 3]
```diff
@@ -118,14 +120,161 @@ public function __construct(array $samlAttributes)
...
+    /**
+     * Get the organization id. This will be the last of the ids given in the claims.
+     *
+     * @return int
+     */
+    public function getOrganizationUnitId(): int
+    {
+        $id = array_slice($this->departmentIds, -1, 1);
+
+        return $id[0];
```
Comment (rimi-itk) [inline Service/SamlDTO.php:162]:
```suggestion
        return end($this->departmentIds);
```

File: Service/SamlDTO.php [hunk 4]
```diff
@@ -118,14 +120,161 @@ public function __construct(array $samlAttributes)
...
+     * @return int
+     *
+     * @throws AakSamlException
+     */
+    public function getMemberOrganizationUnitId(): int
+    {
+        if ($this->isTeamLead()) {
+            $id = array_slice($this->departmentIds, -2, 1);
+        } else {
+            $id = array_slice($this->departmentIds, -1, 1);
```
Comment (rimi-itk) [inline Service/SamlDTO.php:191]: A private `getDepatmentId($offsetFromEnd)` function may come in handy for reducing code duplication of the array slicing (and throwing exceptions).

**Review by rimi-itk — APPROVED** [review]
Approved with some comments and suggestions and questions.

File: Migrations/Version20240701112200.php (hunk 1 above)
Comment (turegjorup) [reply Migrations/Version20240701112200.php:37]: Generated with the `--formatted` option
```
bin/console doctrine:migrations:diff --em=default --configuration=var/plugins/AakSamlBundle/Migrations/aak-saml.yaml --formatted
```

File: Service/SamlDTO.php (hunk 3 above)
Comment (turegjorup) [reply Service/SamlDTO.php:162]:
```
160    Method KimaiPlugin\AakSamlBundle\Service\SamlDTO::getOrganizationUnitId() should return int but returns int|false.  
160    Parameter #1 $array is passed by reference so it does not accept readonly property                                  
         KimaiPlugin\AakSamlBundle\Service\SamlDTO::$departmentIds.
```

**Pull request merged**